Skip Plans step in Sensei onboarding flow if user is coming from SenseiLMS.com

renatho opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #85170

## Proposed Changes

* Auto-select billing period in the plans step in Sensei onboarding flow if the user is coming from a URL with the `plan` param in the URL, and it matches one of the options: "MONTHLY" and "ANNUALLY".
* In parallel, I made 2 changes on senseilms.com:
  * Remove the tooltip about monthly plans from senseilms.com, as I commented in https://github.com/Automattic/wp-calypso/issues/85170#issuecomment-1865145282.
  * Add the `plan` argument to the link "Create A Site" from https://senseilms.com/pricing/.

> I do wonder if it would make sense to create the site earlier in the process—maybe after the domain step—and in the background, so it doesn't block other steps and we can reduce wait time and a loading screen. None of the things [createAndConfigureSite](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/steps-repository/sensei-plan/create-sensei-site.ts#L66C8-L66C30) does is dependent on user input.

The next step after the domain step, is the plans step. So it is now immediately starting the site creation after the domain step when the user is coming from senseilms.com. I thought to make it in parallel with the the domain step, and I started it in https://github.com/Automattic/wp-calypso/pull/85810/commits/4a97317c1ca69d967d82349dbd996c67d96346c7 but I noticed the site creation [depends on the site domain](https://github.com/Automattic/wp-calypso/blob/2bdb35a42e47eb444336b997d5172adabed6a4f1/packages/data-stores/src/onboard/actions.ts#L195), so I reverted it https://github.com/Automattic/wp-calypso/pull/85810/commits/cdd44c158f202a8be8a5570a53f02f1857271aa1. It seems it could use other inputs as fallback, but if we select a domain, and then try to select another one, it would create a site with the previous domain, so it would demand some other changes with the site creation to avoid issues with edge cases then I decided to skip it for now.

What we could try to do in the future, is to run some things from `SenseiLaunch` in parallel with the purpose step. I didn't dig into that to see if that's totally feasible, but I think so. We probably only need to run the `savePurposeData` later.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Navigate to `https://senseilms.com/pricing/`.
* Make sure the Sensei Bundle doesn't have a tooltip talking about monthly prices anymore.
* Copy the link from "Create A Site".
* Replace the domain with `http://calypso.localhost:3000`.
* Go through the steps.
* Make sure you're not asked to choose a plan.
* Make sure the annual billing period is selected.
* Make sure the whole process works properly, and the site is created correctly.
* Now navigate to `http://calypso.localhost:3000/setup/sensei/`.
* Go through the steps, and make sure you're asked to choose a plan.
* Make sure the whole process works properly, and the site is created correctly.
* Test again with `http://calypso.localhost:3000/setup/sensei/?plan=ANYTHING`, and make sure the plan is ignored.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

renatho committed:
Remove not used arguments

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-domain/index.tsx:
```diff
@@ -29,12 +29,12 @@ const SenseiDomain: Step = ( { navigation } ) => {
 
 	const onSkip = () => {
 		setDomain( domain );
-		submit?.( { domain: domain } );
```
This argument wasn't being used anywhere, so I removed it. The same applies to the other `submit` call in this step.

renatho committed:
Auto-select plan if coming from senseilms.com

renatho committed:
Add another check just to avoid edge cases

renatho committed:
Update create site to have a default progress title

renatho committed:
Update auto select to immediately move to bundling status

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-plan/create-sensei-site.ts:
```diff
@@ -54,20 +54,20 @@ export const useCreateSenseiSite = () => {
 	const locale = useLocale();
 	const visibility = useNewSiteVisibility();
 
-	const createAndConfigureSite = useCallback( async () => {
```
The changes on this file are just to make sure we have a default progress title.

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-plan/create-sensei-site.ts:
```diff
@@ -54,20 +54,20 @@ export const useCreateSenseiSite = () => {
 
+	const [ progress, setProgress ] = useState< Progress >( {
+		percentage: 0,
+		title: siteProgressTitle,
```
The changes on this file are just to make sure we have a default progress title.

renatho committed:
Refactor plan selection

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-plan/use-plan-selection.ts:
```diff
@@ -0,0 +1,104 @@
+		}
+	}, [ setStatus, createAndConfigureSite, createCheckoutCart ] );
+
+	// Auto-select YEARLY plan when coming from senseilms.com site.
```
This is the main change of this PR.
The rest of this file is mostly the plan selection logic that was extracted from the component file to simplify it.

renatho committed:
Rename loading variable to make it more clear

renatho committed:
Create site in parallel with the domain selection

renatho committed:
Revert "Create site in parallel with the domain selection"

This reverts commit 4a97317c1ca69d967d82349dbd996c67d96346c7.

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-domain/index.tsx:
```diff
@@ -1,4 +1,3 @@
-/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Removed because it was not necessary here.

Imran92 reviewed (approved):
Looks good and works as expected 👍 Approved!

Not important, but just want to share a thought that I was just wondering if we should add a new param for plans instead of using the ref. That way, if by any chance we add a monthly plan card/panel in senseilmscom, we won't need to update the logic here. Also, if we later redirect here from a different blog in the future, and want a plan to be pre-selected, but with a different ref, we'd be able to use the similar logic.

renatho commented:
> Not important, but just want to share a thought that I was just wondering if we should add a new param for plans instead of using the ref. That way, if by any chance we add a monthly plan card/panel in senseilmscom, we won't need to update the logic here. Also, if we later redirect here from a different blog in the future, and want a plan to be pre-selected, but with a different ref, we'd be able to use the similar logic.

Good point, @Imran92! I'll implement this change before merging it!

renatho committed:
Update logic to be able to set the plan through the URL

renatho commented:
@Imran92, I updated it in this commit: https://github.com/Automattic/wp-calypso/pull/85810/commits/ec5f82f90901f11f09e9aba6468ec112d1862996

And I also updated the PR description and Testing Instructions to match the changes. 😉

Imran92 committed:
Merge branch 'trunk' into change/skip-plan-step-when-coming-from-senseilmscom

Imran92 reviewed (approved):
Looks good and works well! 🥳

donnapep commented:
> * Remove the tooltip about monthly plans from senseilms.com

Looks like there's another one of these when you click on _Compare all features_:

![Screenshot 2024-01-22 at 4 07 05 PM](https://github.com/Automattic/wp-calypso/assets/1190420/d1714fba-354a-47fe-848d-1a1cab7cb2e0)
